### PR TITLE
Support for JSON Smile format for EventReceiverFirehoseFactory

### PR DESCRIPTION
**Pull request by nishantmonu51**
Support for JSON Smile format for EventReceiverFireshoseFactory.
would help in decreasing network load on middle managers ingesting high
data volumes.

### CODE REVIEW

**Comment by drcrallen** [comment]
@nishantmonu51 Do you have data on how much it impacts performance?

File: integration-tests/src/main/java/io/druid/testing/clients/EventReceiverFirehoseTestClient.java [hunk 1]
```diff
@@ -115,6 +132,8 @@ public int postEventsFromFile(String file)
       );
       String s;
       Collection<Map<String, Object>> events = new ArrayList<Map<String, Object>>();
+      // Test sending events using both jsonMapper and smileMapper.
```
Comment (drcrallen) [inline integration-tests/src/main/java/io/druid/testing/clients/EventReceiverFirehoseTestClient.java]: Can you explain this more?

File: integration-tests/src/main/java/io/druid/testing/clients/EventReceiverFirehoseTestClient.java [hunk 2]
```diff
@@ -115,6 +132,8 @@ public int postEventsFromFile(String file)
       );
```
Comment (drcrallen) [inline integration-tests/src/main/java/io/druid/testing/clients/EventReceiverFirehoseTestClient.java]: Should this be try with resources? I don't see the reader closed

File: integration-tests/src/main/java/io/druid/testing/clients/EventReceiverFirehoseTestClient.java [hunk 3]
```diff
@@ -123,12 +142,19 @@ public int postEventsFromFile(String file)
                 }
             )
         );
+        ObjectMapper mapper = (totalEventsPosted % 2 == 0) ? jsonMapper : smileMapper;
+        String mediaType = (totalEventsPosted % 2 == 0)
+                           ? MediaType.APPLICATION_JSON
+                           : SmileMediaTypes.APPLICATION_JACKSON_SMILE;
+        int eventsPosted = postEvents(events, mapper, mediaType);
+        if (eventsPosted != 1) {
```
Comment (drcrallen) [inline integration-tests/src/main/java/io/druid/testing/clients/EventReceiverFirehoseTestClient.java]: Why the change from events.size?

File: server/src/main/java/io/druid/segment/realtime/firehose/EventReceiverFirehoseFactory.java [hunk 4]
```diff
@@ -123,9 +139,30 @@ public EventReceiverFirehose(MapInputRowParser parser)
...
+      Collection<Map<String, Object>> events = null;
+      try {
+        events = objectMapper.readValue(
+            in, new TypeReference<Collection<Map<String, Object>>>()
+            {
+            }
+        );
+      }
+      catch (IOException e) {
+        Throwables.propagate(e);
```
Comment (drcrallen) [inline server/src/main/java/io/druid/segment/realtime/firehose/EventReceiverFirehoseFactory.java]: Can you wrap this? Maybe log locally and return just the message in a map of "error"->e.getMessage

File: server/src/main/java/io/druid/segment/realtime/firehose/EventReceiverFirehoseFactory.java [hunk 5]
```diff
@@ -146,12 +183,18 @@ public Response addAll(Collection<Map<String, Object>> events)
           }
         }
 
-        return Response.ok().entity(ImmutableMap.of("eventCount", events.size())).build();
+        return Response.ok(
+            objectMapper.writeValueAsString(ImmutableMap.of("eventCount", events.size())),
+            contentType
```
Comment (drcrallen) [inline server/src/main/java/io/druid/segment/realtime/firehose/EventReceiverFirehoseFactory.java]: I think there are a lot of other places that do not do this (add content type to the Response). Do we need to add it elsewhere?

Comment (nishantmonu51) [inline server/src/main/java/io/druid/segment/realtime/firehose/EventReceiverFirehoseFactory.java]: the general contract that the existing apis follow is that if the content received is smile, the response is also sent in smile format. e.g. in QueryResource. Its better to set content type if someone relies on that to parse the response.

File: integration-tests/src/main/java/io/druid/testing/clients/EventReceiverFirehoseTestClient.java (hunk 2 above)
Comment (nishantmonu51) [inline integration-tests/src/main/java/io/druid/testing/clients/EventReceiverFirehoseTestClient.java]: fixed.

**Comment by nishantmonu51** [comment]
the major improvement this PR does is improvement in network usage and would help in conserving bandwidth when we run multiple peons in the same machine.

File: integration-tests/src/main/java/io/druid/testing/clients/EventReceiverFirehoseTestClient.java [hunk 6]
```diff
@@ -123,12 +145,19 @@ public int postEventsFromFile(String file)
                 }
             )
         );
+        ObjectMapper mapper = (totalEventsPosted % 2 == 0) ? jsonMapper : smileMapper;
+        String mediaType = (totalEventsPosted % 2 == 0)
+                           ? MediaType.APPLICATION_JSON
+                           : SmileMediaTypes.APPLICATION_JACKSON_SMILE;
+        totalEventsPosted += postEvents(events, mapper, mediaType);
+        ;
```
Comment (gianm) [inline integration-tests/src/main/java/io/druid/testing/clients/EventReceiverFirehoseTestClient.java]: extraneous semicolon

**Comment by gianm** [comment]
:+1: lgtm. Compression could be cool too in the future, as I recall from testing things in Kafka that smile-lz4 was a good combination of cpu and size.
